File: python/structural/decorator.py
import time


def time_decorator(fn):
    def inner_fun(*args, **kwargs):
        start = time.time()
        result = fn(*args, **kwargs)
        end = time.time()
        duration = (end - start)
        print(f"function {fn.__name__} executed in {duration:4f}")
        return result
    return inner_fun

@time_decorator
def slow_func():
    time.sleep(1.5)

slow_func()

File: js/structural/decorator.js
const testFunc = () => {
    for (let i = 0; i < 9999; i++) {
        const num = Math.random() * 2 / 55 * 1000 + i
        console.log(num)
    }
}


const timeDecorator = (fn) => {
    return (...args) => {
        const start = new Date()
        const result = fn(...args)
        const end = new Date()
        const duration = (end - start).toFixed(4)
        console.log(`function ${fn.name} executed in ${duration}`)
        return result
    }

}

const decoratedFunc = timeDecorator(testFunc)
decoratedFunc()